### DISCUSSION TITLE
Conflict with Node.js Core Modules in Next.js due to Webpack Fallback Configuration

### DISCUSSION THREAD

**Discussion by Ad-h0c** [vercel/next.js#58642] [Help] [answered]
### Summary

I'm encountering an issue in my Next.js project related to the use of Node.js core modules (`fs`, `path`, `os`) in a `node-gyp` file within my `node_modules`. My `next.config.js` is configured with custom Webpack settings that include `config.resolve.fallback = { fs: false, net: false, tls: false }`. This setting is intended to prevent Webpack from polyfilling these Node.js-specific modules for browser use.

However, the `node-gyp` file (and possibly other dependencies) relies on these core modules (`fs`, `path`, `os`), which are native to Node.js and not available in the browser environment. Since my Webpack configuration explicitly instructs not to polyfill `fs`, I am concerned this might be causing compatibility issues or runtime errors, particularly when this code or its dependencies might be included in the client-side bundle.

I'm looking for guidance on how to best handle this situation. Specifically, I need to know if there's a way to ensure that `node-gyp` and its dependencies are only used server-side, or if there's an alternative approach to configure Webpack in my Next.js project to avoid such conflicts without compromising the functionality of my application.

Import trace for requested module:
./node_modules/node-gyp-build/index.js
./node_modules/bufferutil/index.js
./node_modules/ws/lib/buffer-util.js
./node_modules/ws/lib/websocket.js
./node_modules/ws/index.js
./node_modules/isows/_esm/index.js
./node_modules/viem/_esm/utils/rpc.js
./node_modules/viem/_esm/clients/transports/http.js
./node_modules/viem/_esm/index.js
./node_modules/@rainbow-me/rainbowkit/dist/chunk-VCSNNDJA.js
./node_modules/@rainbow-me/rainbowkit/dist/index.js
./app/provider.tsx
 ⚠ ./node_modules/node-gyp-build/index.js
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

Import trace for requested module:
./node_modules/node-gyp-build/index.js
./node_modules/bufferutil/index.js
./node_modules/ws/lib/buffer-util.js
./node_modules/ws/lib/websocket.js
./node_modules/ws/index.js
./node_modules/isows/_esm/index.js
./node_modules/viem/_esm/utils/rpc.js
./node_modules/viem/_esm/clients/transports/http.js
./node_modules/viem/_esm/index.js
./node_modules/@rainbow-me/rainbowkit/dist/chunk-VCSNNDJA.js
./node_modules/@rainbow-me/rainbowkit/dist/index.js
./app/provider.tsx
 ⚠ ./node_modules/node-gyp-build/index.js
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

Import trace for requested module:
./node_modules/node-gyp-build/index.js
./node_modules/bufferutil/index.js
./node_modules/ws/lib/buffer-util.js
./node_modules/ws/lib/websocket.js
./node_modules/ws/index.js
./node_modules/isows/_esm/index.js
./node_modules/viem/_esm/utils/rpc.js
./node_modules/viem/_esm/clients/transports/http.js
./node_modules/viem/_esm/index.js
./node_modules/@rainbow-me/rainbowkit/dist/chunk-VCSNNDJA.js
./node_modules/@rainbow-me/rainbowkit/dist/index.js
./app/provider.tsx
 ⚠ ./node_modules/node-gyp-build/index.js
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

Import trace for requested module:
./node_modules/node-gyp-build/index.js
./node_modules/bufferutil/index.js
./node_modules/ws/lib/buffer-util.js
./node_modules/ws/lib/websocket.js
./node_modules/ws/index.js
./node_modules/isows/_esm/index.js
./node_modules/viem/_esm/utils/rpc.js
./node_modules/viem/_esm/clients/transports/http.js
./node_modules/viem/_esm/index.js
./node_modules/@rainbow-me/rainbowkit/dist/chunk-VCSNNDJA.js
./node_modules/@rainbow-me/rainbowkit/dist/index.js
./app/provider.tsx

Next config file:

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack: config => {
    config.resolve.fallback = { fs: false, net: false, tls: false };
    config.externals.push('pino-pretty', 'lokijs', 'encoding');
    return config;
  },
};

module.exports = nextConfig
```

### Additional information

```js
Operating System:
  Platform: linux
  Arch: x64
  Version: #37~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon Oct  9 15:34:04 UTC 2
Binaries:
  Node: 18.18.0
  npm: 9.8.1
  Yarn: 1.22.15
  pnpm: 6.11.0
Relevant Packages:
  next: 14.0.3
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

**Comment by icyJoseph** [comment]
Try this:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack: (config, { nextRuntime }) => {
    if (typeof nextRuntime === "undefined") {
      config.resolve.fallback = {
        ...config.resolve.fallback,
        fs: false,
        net: false,
        tls: false,
      };
    }
    config.externals.push("pino-pretty", "lokijs", "encoding");
    return config;
  },
};

module.exports = nextConfig;
```

The `nextRuntime` can be `nodejs`, `edge` or `undefined`, which would be for the browser. This way you only apply these fallbacks for browser bundles. You might have to tweak it a bit, not sure about the externals you have.

**Comment by Ad-h0c** [comment, marked as answer]
```
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack: config => {
    config.resolve.fallback = { fs: false, net: false, tls: false };
    config.externals.push('pino-pretty', 'lokijs', 'encoding', { 'node-gyp-build': 'commonjs node-gyp-build' });
    return config;
  },
};

module.exports = nextConfig
```

So I told Webpack to treat node-gyp-build as an external module and to use the CommonJS module system when requiring it. And that solved the problem.
